File: api/app/routers/digest.py
from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query, Security
from sqlalchemy.exc import IntegrityError
from sqlalchemy.ext.asyncio import AsyncSession
from starlette.requests import Request

from app.database import get_session
from app.dependencies import verify_api_key
from app.rate_limit import limiter
from app.schemas.digest import (
    DigestCreate,
    DigestListResponse,
    DigestResponse,
)
from app.services.digest_service import (
    create_digest,
    get_digest_by_date,
    get_digests,
)

router = APIRouter(prefix="/digest", tags=["digest"])


@router.post("", response_model=DigestResponse, status_code=201)
@limiter.limit("5/minute")
async def create_digest_endpoint(
    request: Request,
    data: DigestCreate,
    session: AsyncSession = Depends(get_session),
    _api_key: str = Security(verify_api_key),
):
    try:
        digest = await create_digest(session, data)
        return digest
    except IntegrityError:
        raise HTTPException(
            status_code=409, detail="Digest for this date already exists"
        )


@router.get("", response_model=DigestListResponse)
@limiter.limit("60/minute")
async def list_digests(
    request: Request,
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    session: AsyncSession = Depends(get_session),
):
    digests, total = await get_digests(session, page, per_page)
    return DigestListResponse(items=digests, total=total, page=page, per_page=per_page)


@router.get("/{digest_date}", response_model=DigestResponse)
@limiter.limit("60/minute")
async def get_digest(
    request: Request,
    digest_date: date,
    session: AsyncSession = Depends(get_session),
):
    digest = await get_digest_by_date(session, digest_date)
    if digest is None:
        raise HTTPException(status_code=404, detail="Digest not found")
    return digest

File: api/src/routes/digest.ts
import { Hono } from "hono";
import { describeRoute, validator } from "hono-openapi";
import type { DB } from "../database.js";
import { jstYesterday } from "../lib/jst-date.js";
import {
  formatDigestListItem,
  formatDigestResponse,
  formatDigestSourceArticle,
} from "../mappers/digest.js";
import { PG_UNIQUE_VIOLATION, getPgErrorCode } from "../middleware/error-handler.js";
import { writeGuard } from "../middleware/guards.js";
import { createRateLimiter } from "../middleware/rate-limit.js";
import { validationHook } from "../middleware/validation.js";
import { API_KEY_SECURITY, errorResponse, jsonResponse } from "../openapi.js";
import { digestDateParamSchema } from "../schemas/base.js";
import {
  type DigestSourceResponse,
  digestCreateSchema,
  digestListQuerySchema,
  digestListResponseSchema,
  digestResponseSchema,
  digestSourceQuerySchema,
  digestSourceResponseSchema,
} from "../schemas/digest.js";
import { getArticlesForDigest } from "../services/article-service.js";
import { createDigest, getDigestByDate, getDigests } from "../services/digest-service.js";
import type { AppEnv } from "../types.js";

const TAGS = ["Digests"];

export function createDigestRoute(db: DB): Hono<AppEnv> {
  const route = new Hono<AppEnv>();

  // POST /digest
  route.post(
    "/digest",
    ...writeGuard(5),
    describeRoute({
      tags: TAGS,
      summary: "Store a daily digest",
      security: API_KEY_SECURITY,
      responses: {
        201: jsonResponse("Created digest", digestResponseSchema),
        401: errorResponse("Missing or invalid API key"),
        409: errorResponse("A digest for this date already exists"),
      },
    }),
    validator("json", digestCreateSchema, validationHook),
    async (c) => {
      const data = c.req.valid("json");
      try {
        const digest = await createDigest(db, data);
        return c.json(formatDigestResponse(digest), 201);
      } catch (err) {
        if (getPgErrorCode(err) === PG_UNIQUE_VIOLATION) {
          return c.json({ detail: "Digest for this date already exists" }, 409);
        }
        throw err;
      }
    },
  );

  // GET /digest
  route.get(
    "/digest",
    createRateLimiter(60),
    describeRoute({
      tags: TAGS,
      summary: "List digests, newest first",
      responses: { 200: jsonResponse("Paginated digests", digestListResponseSchema) },
    }),
    validator("query", digestListQuerySchema, validationHook),
    async (c) => {
      const { page, per_page } = c.req.valid("query");
      const { items, total } = await getDigests(db, page, per_page);
      return c.json({
        items: items.map(formatDigestListItem),
        total,
        page,
        per_page,
      });
    },
  );

  // GET /digest/source-articles — full source content for a JST day, for LLM digest generation.
  // Registered before /digest/:digest_date so the static segment is not parsed as a date.
  route.get(
    "/digest/source-articles",
    createRateLimiter(60),
    describeRoute({
      tags: TAGS,
      summary: "Full article bodies created on one JST day (defaults to yesterday)",
      responses: { 200: jsonResponse("Source articles", digestSourceResponseSchema) },
    }),
    validator("query", digestSourceQuerySchema, validationHook),
    async (c) => {
      const { date } = c.req.valid("query");
      const targetDate = date ?? jstYesterday(new Date());
      const { items, truncated } = await getArticlesForDigest(db, targetDate);
      const response: DigestSourceResponse = {
        date: targetDate,
        count: items.length,
        truncated,
        articles: items.map(formatDigestSourceArticle),
      };
      return c.json(response);
    },
  );

  // GET /digest/:digest_date
  route.get(
    "/digest/:digest_date",
    createRateLimiter(60),
    describeRoute({
      tags: TAGS,
      summary: "Digest for a date (YYYY-MM-DD)",
      responses: {
        200: jsonResponse("Digest", digestResponseSchema),
        404: errorResponse("Digest not found"),
      },
    }),
    validator("param", digestDateParamSchema, validationHook),
    async (c) => {
      const { digest_date: digestDate } = c.req.valid("param");
      const digest = await getDigestByDate(db, digestDate);
      if (!digest) {
        return c.json({ detail: "Digest not found" }, 404);
      }
      return c.json(formatDigestResponse(digest));
    },
  );

  return route;
}
